### ISSUE TITLE
Doesn't seemingly work with latest TF versions.

### ISSUE THREAD

**Issue by joeyearsley** [openai/gradient-checkpointing#29]
Having successfully used this when it came out with TF 1.5 it seemingly doesn't work anymore in TF 1.9.

@yaroslavvb Do you have a working version still? Or do you have any insights as to what might have changed?

**Comment by joeyearsley** [comment]
No explicit error, just doesn’t seemingly work like in 1.5 .

**Comment by varun19299** [comment]
Nope this doesn't seem to work even in 1.11; even if graph optimisations are turned off.

**Comment by stefano-marchesin** [comment]
Not woking in 1.11, it works smooth for a while and then (randomly) throws OOM exception.

**Comment by joeyearsley** [comment]
As a work around I’ve been using `tf.contrib.layers.recompute_grad` in predefined places.

**Comment by ekyy2** [comment]
I find that the tests (`./run_all_tests.sh`) work fine for TensorFlow 1.8, but break for versions 1.9 to 1.12 with the following error:
> Traceback (most recent call last):
>   File "./keras_test.py", line 4, in <module>
>     from tensorflow.python.keras._impl.keras import backend as K
> ModuleNotFoundError: No module named 'tensorflow.python.keras._impl'

**Comment by racinmat** [comment]
Today I compiled the 1.12 version from sources and this library works well. 
It disappointed me a bit, because the 'memory'  option failed. I use it for CycleGAN (batch size 1), and it failed to find any bottleneck. 
When using the 'speed' option, it improved a little bit, but not much around 1-2%. 

@ekyy2  this is just because of moving packages, if you replace the import by `from tensorflow.python.keras import backend as K` it will work as charm and you can run the tests on 1.12 without problem.

**Comment by joeyearsley** [comment]
If using TF you can manually wrap layers like I mention above you can make larger memory gains but your files per second will drop if done inefficiently - probably matters less if you need more memory.

**Comment by racinmat** [comment]
@joeyearsley  did the `tf.contrib.layers.recompute_grad` help you with the memory management? Do you have some examples how well it performed when placed correctly to the code?

**Comment by joeyearsley** [comment]
Here is an example with Densenet:
https://github.com/joeyearsley/efficient_densenet_tensorflow

Be wary when using with dropout though, you'll need to implement a new dropout layer which takes an `is_recomputing` kwarg which stores the random vector created in the forward pass to be used in the backward pass also. Otherwise the gradients will not be correct.

**Comment by ekyy2** [comment]
> 
> 
> Today I compiled the 1.12 version from sources and this library works well.
> It disappointed me a bit, because the 'memory' option failed. I use it for CycleGAN (batch size 1), and it failed to find any bottleneck.
> When using the 'speed' option, it was even slower than the original gradient calculation, by 10%.
> 
> @ekyy2 this is just because of moving packages, if you replace the import by `from tensorflow.python.keras import backend as K` it will work as charm and you can run the tests on 1.12 without problem.



> 
> 
> Today I compiled the 1.12 version from sources and this library works well.
> It disappointed me a bit, because the 'memory' option failed. I use it for CycleGAN (batch size 1), and it failed to find any bottleneck.
> When using the 'speed' option, it was even slower than the original gradient calculation, by 10%.
> 
> @ekyy2 this is just because of moving packages, if you replace the import by `from tensorflow.python.keras import backend as K` it will work as charm and you can run the tests on 1.12 without problem.

That did the trick!

**Comment by ImranRolo** [comment]
> Here is an example with Densenet:
> https://github.com/joeyearsley/efficient_densenet_tensorflow
> 
> Be wary when using with dropout though, you'll need to implement a new dropout layer which takes an `is_recomputing` kwarg which stores the random vector created in the forward pass to be used in the backward pass also. Otherwise the gradients will not be correct.

Hi Joey, how do you use `tf.contrib.layers.recompute_grad` in conjunction  with `memory_saving_gradients.gradients(...)`. From your code, it looks like you only used `recompute_grad`.

**Comment by joeyearsley** [comment]
Yes I stopped using `memory_saving_gradients` As recompute grad works just as well but with more manual work.

**Comment by ImranRolo** [comment]
> Yes I stopped using `memory_saving_gradients` As recompute grad works just as well but with more manual work.

Thanks for your reply Joey. I can see that you recompute the gradient. I'm just slightly lost as to where in the code you drop gradients from memory as well.

**Comment by yaroslavvb** [comment]
A tensor is dropped automatically as long as it no longer has downstream consumers, this is what [checkpoints_disconnected](https://github.com/openai/gradient-checkpointing/blob/master/memory_saving_gradients.py#L189) is for -- because of stop_gradient it doesn't need to keep activations in memory, and then `reroute_ts` substitutes those disconnected tensors in place of originals to drop the activations

**Comment by ImranRolo** [comment]
> dropped automatically

Thanks yaroslavvb. Unfortunately, I have not found `memory_saving_gradients` to work for tensorflow1.12 like others above. I'm hoping `tf.contrib.layers.recompute_grad` works.

**Comment by gitrdonator** [comment]
When using `recompute_grad`, are the weights completely lost? I have no trainable parameters on the layers where I use it.

**Comment by gitrdonator** [comment]
I can confirm @yaroslavvb 's `memory_saving_gradients` are working in TensorFlow 1.14.0. I simply needed to turn off optimizations as he suggested.

**Comment by gitrdonator** [comment]
I confirmed too soon. My training was behaving very strangely, so I commented out `K.__dict__['gradients'] = memory_saving_gradients.gradients_memory`. The model still runs fast and with little memory, which means the memory saving and speed came from @yaroslavvb 's optimizations, but not `memory_saving_gradients`. I'm back at square one...it doesn't seem like `memory_saving_gradients` is working.

**Comment by gitrdonator** [comment]
Same issue. `memory_saving_gradients` does not work in tf 1.14.

**Comment by golden0080** [comment]
Tested in TF 1.15, using Keras. No luck for me.
I got some errors like:
```
tensorflow.python.framework.errors_impl.InvalidArgumentError: Operation 'training/Adam/gradients/header/conv_4/batch_normalization_63/cond/ReadVariableOp_2/Switch' has no attr named '_XlaCompile'.
```
